ddl/index_merge_tmp.go

```diff
@@ -0,0 +1,238 @@
+	ctx := kv.WithInternalSourceType(context.Background(), w.jobContext.ddlJobSourceType())
+	errInTxn = kv.RunInNewTxn(ctx, w.sessCtx.GetStore(), true, func(ctx context.Context, txn kv.Transaction) error {
+		taskCtx.addedCount = 0
+		taskCtx.scanCount = 0
+		txn.SetOption(kv.Priority, w.priority)
+		if tagger := w.reorgInfo.d.getResourceGroupTaggerForTopSQL(w.reorgInfo.Job); tagger != nil {
+			txn.SetOption(kv.ResourceGroupTagger, tagger)
+		}
+
+		tmpIdxRecords, nextKey, taskDone, err := w.fetchTempIndexVals(txn, taskRange)
+		if err != nil {
+			return errors.Trace(err)
+		}
+		taskCtx.nextKey = nextKey
+		taskCtx.done = taskDone
+
+		err = w.batchCheckTemporaryUniqueKey(txn, tmpIdxRecords)
+		if err != nil {
+			return errors.Trace(err)
+		}
+
+		for i, idxRecord := range tmpIdxRecords {
+			// The index is already exists, we skip it, no needs to backfill it.
+			// The following update, delete, insert on these rows, TiDB can handle it correctly.
+			// If all batch are skipped, update first index key to make txn commit to release lock.
+			if idxRecord.skip {
+				continue
+			}
+			if idxRecord.delete {
+				if idxRecord.unique {
+					err = txn.GetMemBuffer().DeleteWithFlags(w.originIdxKeys[i], kv.SetNeedLocked)
+				} else {
+					err = txn.GetMemBuffer().Delete(w.originIdxKeys[i])
+				}
+			} else {
+				err = txn.GetMemBuffer().Set(w.originIdxKeys[i], idxRecord.vals)
+			}
+			if err != nil {
+				return err
+			}
+		}
+		return nil
+	})
+	logSlowOperations(time.Since(oprStartTime), "AddIndexMergeDataInTxn", 3000)
+	return
+}
+
+func (w *mergeIndexWorker) AddMetricInfo(cnt float64) {
+}
+
+func (w *mergeIndexWorker) fetchTempIndexVals(txn kv.Transaction, taskRange reorgBackfillTask) ([]*temporaryIndexRecord, kv.Key, bool, error) {
+	startTime := time.Now()
+	w.tmpIdxRecords = w.tmpIdxRecords[:0]
+	w.tmpIdxKeys = w.tmpIdxKeys[:0]
+	w.originIdxKeys = w.originIdxKeys[:0]
+	// taskDone means that the merged handle is out of taskRange.endHandle.
+	taskDone := false
+	oprStartTime := startTime
+	idxPrefix := w.table.IndexPrefix()
+	var lastKey kv.Key
+	err := iterateSnapshotKeys(w.reorgInfo.d.jobContext(w.reorgInfo.Job), w.sessCtx.GetStore(), w.priority, idxPrefix, txn.StartTS(),
+		taskRange.startKey, taskRange.endKey, func(_ kv.Handle, indexKey kv.Key, rawValue []byte) (more bool, err error) {
+			oprEndTime := time.Now()
+			logSlowOperations(oprEndTime.Sub(oprStartTime), "iterate temporary index in merge process", 0)
+			oprStartTime = oprEndTime
+
+			if taskRange.endInclude {
+				taskDone = indexKey.Cmp(taskRange.endKey) > 0
+			} else {
+				taskDone = indexKey.Cmp(taskRange.endKey) >= 0
+			}
+
+			if taskDone || len(w.tmpIdxRecords) >= w.batchCnt {
+				return false, nil
+			}
+
+			isDelete := false
+			unique := false
+			var keyVer []byte
+			length := len(rawValue)
+			keyVer = append(keyVer, rawValue[length-1:]...)
+			rawValue = rawValue[:length-1]
+			length--
+			if bytes.Equal(keyVer, []byte("m")) {
```

Comment:
Do we need to use 'tempKeyVerMerge' here?

Reply:
Done.